### ISSUE TITLE
Proposal: method to include predefined query function in query chain

### ISSUE THREAD

**Issue by danieldunderfelt** [rethinkdb/rethinkdb#5838]
Reusing queries is a good pattern to keep things DRY, but ReQL doesn't have a smooth way of doing so. Take this query fragment for example:

```javascript
function mergeCommentCount(query) {
  return query.merge(row => ({
    commentsCount: r.table('Comment').count(comment => {
      return comment('postId').eq(row('id'))
    })
  }))
}
```
We'd like to reuse it in every query that needs to have a `commentsCount` field, so we do this:

```javascript
return mergeCommentCount(r.table('Post')).run()
```
A one-liner isn't so bad, but what if you want to chain more queries onto it? If you have more than one query fragment function, you'd need to create a temporary variable to hold the query for each reusable query function.

**My proposal:**

`.include(queryFunction)`

Having a ReQL method that takes a callback that you can chain into the rest of the queries would be ideal. For example:

```javascript
// Random ReQL
r.table('Post')
  .filter(...)
  .include(queryFunc1)
  .orderBy(...)
  .include(queryFunc2)
  .include(queryFunc3)
  .slice(...)
  .run()
```

Much cleaner, yes? `include` would simply call the callback with the current query so far and include the resulting query returned by the queryFunc into the query chain. It would be up to the user to call the predefined queries in a supported order.

Thoughts?

**Comment by mglukhovsky** [comment]
Underscore.js has something similar, called [`tap`](http://underscorejs.org/#tap):

> Invokes interceptor with the object, and then returns object. The primary purpose of this method is to "tap into" a method chain, in order to perform operations on intermediate results within the chain.

**Comment by deontologician** [comment]
tap returns `this`, whereas include returns `f(this)`

**Comment by mglukhovsky** [comment]
Right. I see why @danielmewes prefers `mapQuery`, but I think there's a better name out there.

**Comment by VeXocide** [comment]
Just throwing a few names out there, how about `chain`, or `dot`?

**Comment by danielmewes** [comment]
I like `chain`.

Also note that this function will not operate on intermediate query results, but on the query object itself.

**Comment by deontologician** [comment]
I guess another question is whether this should actually be a reql function applied on the server. As it is, this is like a reql macro, and errors reported by the server, stack traces etc, will refer to the expanded version if we do the function application client side.

**Comment by danielmewes** [comment]
I assumed this would be client-side, mostly because it's way easier to implement.

If we want to do this server-side, we'll need an additional evaluation step on the server I think, or some other significant change to how queries are executed (because there's no longer just a single "compilation" pass). I could be over-estimating the amount of work required though.

We could consider allowing the client to add "annotations" to terms that describe which `chain` command the terms originated from. Then the server could take those annotations into account when generating backtraces. I don't think this would actually be helpful though, as it would also make it harder to find out where exactly inside a query template the error occurred.

**Comment by mlucy** [comment]
+1 for adding this and doing it client-side.  Also, since it wasn't explicitly stated above, it should probably be variadic, so you can write `reql_query.coolname(f, arg2, arg3)` to get `f(reql_query, arg2, arg3)`.

As far as the name, I would vote for `chain` or `funcall`.

**Comment by danieldunderfelt** [comment]
@mlucy +1 for variadic!

I also like `chain` more than `include`, which was really just a descriptive placeholder. Fun to see this get interest!

**Comment by larkost** [comment]
I had a proposal for this a while ago: #4576. It basically inverts the idea, rather than having an `.include()` function, it has a `r.fragment` that you can prepend onto your pre-defined part when assigning it to a variable. Then you just put that variable into your query/queries. This of course would all be done client-side, and the server would never know that you had done anything like this.

**Comment by deoqc** [comment]
Just like [d3js call](https://github.com/d3/d3-selection/blob/master/README.md#selection_call), which is quite a compliment.
